python/cudf/cudf/pandas/fast_slow_proxy.py

```diff
@@ -898,6 +899,18 @@ def _fast_slow_function_call(func: Callable, /, *args, **kwargs) -> Any:
                 # try slow path
                 raise Exception()
             fast = True
+
+            if get_option("mode.pandas_debugging"):
+                with nvtx.annotate(
+                    "EXECUTE_SLOW",
+                    color=_CUDF_PANDAS_NVTX_COLORS["EXECUTE_SLOW"],
+                    domain="cudf_pandas",
+                ):
+                    slow_args, slow_kwargs = _slow_arg(args), _slow_arg(kwargs)
+                    with disable_module_accelerator():
+                        result_slow = func(*slow_args, **slow_kwargs)
+                        print(result_slow)
+                        # TODO: Compare result and result slow and return a warning
```

Comment:
How should I go about checking if `result` (the fast one) and `result_slow` are equal? 

The function `_fast_slow_function_call` is called multiple times for a given piece of pandas code. In the example below, I added some print statements with "FAST" and "SLOW" and the type of the `result` and `result_slow`, respectively. I could do the check when it makes sense to (eg. when the types are the same) or when they have the same meaning like a cudf DataFrame and a pandas DataFrame.

```python
rankings = {
  "name": ["SCOTTIE SCHEFFLER", "XANDER SCHAUFFELE", "RORY MCILROY", "WYNDHAM CLARK", "VIKTOR HOVLAND"],
  "avg_points": [10, 9, 8 ,7 ,6]
}

df = pd.DataFrame(rankings)

print(df) 
```

```
FAST  <class 'NoneType'>
SLOW  <class 'NoneType'>
FAST  <class 'cudf.core.dataframe._DataFrameLocIndexer'>
SLOW  <class 'pandas.core.indexing._LocIndexer'>
FAST  <class 'cudf.core.dataframe.DataFrame'>
SLOW  <class 'pandas.core.frame.DataFrame'>
FAST  <class 'str'>
SLOW  <class 'str'>
                name  avg_points
0  SCOTTIE SCHEFFLER          10
1  XANDER SCHAUFFELE           9
```

Reply:
You'll probably want to have a helper function that calls [pandas assertion functions](https://pandas.pydata.org/docs/reference/testing.html#assertion-functions) if this is set.

```
def compare_fast_slow(fast_result, slow_result):
     if isinstance(fast_result, cudf.DataFrame):
         tm.assert_frame_equal(fast_result.to_pandas(), slow_result)
     elif isinstance(fast_result, cudf.Series):
         tm.assert_series_equal(fast_result.to_pandas(), slow_result)
     ...
     else:
         assert fast_result == slow_result
```

Also to reduce code duplication, you can make this branch `raise Exception()` so the slow path below runs and you can do this comparison in that block

Reply:
Should we raise warnings instead of exception? There are some cases where we expect the results to disagree like joins and groupbys. I'm not sure which cases we should handle in `compare_fast_slow`. For example, should we also have a case for `cudf.core.dataframe._DataFrameLocIndexer` and `pandas.core.indexing._LocIndexer`?

This is supposed to be a debugging tool so more case would help us pinpoint where the issues are in cudf. But there are also concerns of verbosity.

Reply:
Should we use [assert_eq](https://github.com/rapidsai/cudf/blob/1710e11c3ae9dd072305ca49e12e10d0f2e3aec0/python/cudf/cudf/testing/_utils.py#L116) here?

Reply:
Yes, I would use assertion functions and catch `AssertionError` -- then raise a warning there.

Also we should issue a different warning if the pandas "slow path" raises an error, and show the content of that error.

Reply:
> Should we use [assert_eq](https://github.com/rapidsai/cudf/blob/1710e11c3ae9dd072305ca49e12e10d0f2e3aec0/python/cudf/cudf/testing/_utils.py#L116) here?

Ah yes that would probably be more agnostic that using the pandas testing functions

> We might do this debugging comparison for only "final" types

Yeah agreed, with @Matt711's concerns about being too verbose comparing everything, I think for debugging we would only care about these final types

Reply:
> Also we should issue a different warning if the pandas "slow path" raises an error, and show the content of that error.

@bdice Do you mean the actual slow path or the slow path in the fast path?

Reply:
Just the slow path in the debug mode of the fast path. We don’t return that result to the user since it’s for the debug comparison — but we should warn if it cannot be computed for some reason.

Reply:
Which kind of exception should we catch if the result cannot be computed? I'm looking at the Exceptions listed on Pandas testing [docs](https://pandas.pydata.org/docs/reference/testing.html). @bdice

Reply:
You want something like:

```python
try:
    compute debug slow path
except Exception:
    warn that computing debug slow path failed
else:
    try:
        assert_eq(…)
    except AssertionError:
         warn that cudf/pandas gave different results
    except Exception:
         warn that equality check failed
```

Reply:
Also we need to figure out how to make this generic: the proxy is meant to be general and not mention cudf/pandas by name (you can use the name attributes of the fast/slow modules, perhaps?). The assertion function and debug mode should also be generic, perhaps those should be arguments.